src/lxc/conf.c

```diff
@@ -1455,48 +1455,92 @@ static int setup_pivot_root(const struct lxc_rootfs *rootfs)
+		      "devices are requested.");
 		return 0;
-
-	if (!access("/dev/pts/ptmx", F_OK) && umount("/dev/pts")) {
-		SYSERROR("failed to umount 'dev/pts'");
-		return -1;
 	}
 
-	if (mkdir("/dev/pts", 0755)) {
-		if ( errno != EEXIST ) {
-		    SYSERROR("failed to create '/dev/pts'");
-		    return -1;
+	ret = access("/dev/pts/ptmx", F_OK);
+	if (!ret) {
+		/* Unmount old devpts instance. */
+		ret = umount("/dev/pts");
+		if (ret < 0) {
+			SYSERROR("Failed to unmount old devpts instance.");
+			return -1;
 		}
+		DEBUG("Unmounted old /dev/pts instance.");
 	}
 
-	if (mount("devpts", "/dev/pts", "devpts", MS_MGC_VAL,
-		  "newinstance,ptmxmode=0666,mode=0620,gid=5")) {
-		SYSERROR("failed to mount a new instance of '/dev/pts'");
+	/* Create mountpoint for devpts instance. */
+	ret = mkdir("/dev/pts", 0755);
+	if (ret < 0 && errno != EEXIST) {
+		SYSERROR("Failed to create the \"/dev/pts\" directory.");
 		return -1;
 	}
 
-	if (access("/dev/ptmx", F_OK)) {
-		if (!symlink("/dev/pts/ptmx", "/dev/ptmx"))
-			goto out;
-		SYSERROR("failed to symlink '/dev/pts/ptmx'->'/dev/ptmx'");
+	/* Mount new devpts instance. */
+	ret = mount("devpts", "/dev/pts", "devpts", MS_MGC_VAL, devpts_mntopts);
+	if (ret < 0) {
+		SYSERROR("Failed to mount new devpts instance.");
 		return -1;
 	}
 
-	if (realpath("/dev/ptmx", target) && !strcmp(target, "/dev/pts/ptmx"))
-		goto out;
+	ret = access("/dev/ptmx", F_OK);
+	if (ret < 0) {
+		/* Create dummy /dev/ptmx file as bind mount target for
+		 * /dev/pts/ptmx.
+		 */
+		ret = open("/dev/ptmx", O_CREAT, 0666);
+		if (ret < 0) {
+			SYSERROR("Failed to create dummy \"/dev/ptmx\" file as bind mount target.");
+			return -1;
+		}
+		created_ptmx = true;
+		DEBUG("Created dummy \"/dev/ptmx\" file as bind mount target.");
+	}
 
-	/* fallback here, /dev/pts/ptmx exists just mount bind */
-	if (mount("/dev/pts/ptmx", "/dev/ptmx", "none", MS_BIND, 0)) {
-		SYSERROR("mount failed '/dev/pts/ptmx'->'/dev/ptmx'");
+	/* Bind mount /dev/pts/ptmx to /dev/ptmx. */
+	ret = mount("/dev/pts/ptmx", "/dev/ptmx", "none", MS_BIND, 0);
+	if (ret < 0) {
+		ERROR("Failed to bind mount \"/dev/pts/ptmx\" to \"/dev/ptmx\"");
+		/* Fallthrough and try to create a symlink. */
+		if (created_ptmx) {
+			ret = remove("/dev/pts");
+			if (ret < 0) {
+				SYSERROR("Failed to remove dummy \"/dev/ptmx\" file.");
+				return -1;
+			}
+		}
+	} else {
+		DEBUG("Bind mounted \"/dev/pts/ptmx\" to \"/dev/ptmx\"");
+		goto success;
+	}
+
+	/* Fallback option: Create symlink /dev/ptmx -> /dev/pts/ptmx. */
+	ret = access("/dev/ptmx", F_OK);
+	if (ret < 0) {
+		ret = symlink("/dev/pts/ptmx", "/dev/ptmx");
+		if (!ret) {
+			DEBUG("Created symlink \"/dev/ptmx\" -> \"/dev/pts/ptmx\".");
+			goto success;
+		}
+		SYSERROR("Failed to create symlink \"/dev/ptmx\" -> \"/dev/pts/ptmx\"");
 		return -1;
 	}
 
-	INFO("created new pts instance");
+	/* Check if the symlink is valid. */
+	if (realpath("/dev/ptmx", target) && !strcmp(target, "/dev/pts/ptmx"))
+		goto success;
```

Comment:
Nah, that's why I'm re-doing the commits and also to split out *all* non-functional changes. :) Coming up in a bit. :)